Mathlib/RingTheory/TwoSidedIdeal/Operations.lean

```diff
@@ -0,0 +1,207 @@
+  written as a finite sum of the form `∑ i, xL i * y i * xR i` where `xL i ∈ s`, `y i ∈ R`, and
+  `xR i ∈ s`.
+
+- `TwoSidedIdeal.comap`: pullback of a two-sided-ideal; defined as the preimage of a
+  two-sided-ideal.
+- `TwoSidedIdeal.map`: pushforward of a two-sided-ideal; defined as the span of the image of a
+  two-sided-ideal.
+- `TwoSidedIdeal.ker`: the kernel of a ring homomorphism as a two-sided-ideal.
+-/
+
+namespace TwoSidedIdeal
+
+section NonUnitalNonAssocRing
+
+variable {R : Type*} [NonUnitalNonAssocRing R]
+
+/--
+The smallest two-sided-ideal containing a set.
+-/
+abbrev span (s : Set R) : TwoSidedIdeal R :=
+  { ringCon := ringConGen (fun a b ↦ a - b ∈ s) }
+
+lemma subset_span {s : Set R} : s ⊆ (span s : Set R) := by
+  intro x hx
+  rw [SetLike.mem_coe, mem_iff]
+  exact RingConGen.Rel.of _ _ (by simpa using hx)
+
+lemma mem_span_iff {s : Set R} {x} :
+    x ∈ span s ↔ ∀ (I : TwoSidedIdeal R), s ⊆ I → x ∈ I := by
+  refine ⟨?_, fun h => h _ subset_span⟩
+  delta span
+  rw [RingCon.ringConGen_eq]
+  intro h I hI
+  refine sInf_le (α := RingCon R) ?_ h
+  intro x y hxy
+  specialize hI hxy
+  rwa [SetLike.mem_coe, ← rel_iff] at hI
+
+lemma span_mono {s t : Set R} (h : s ⊆ t) : span s ≤ span t := by
+  intro x hx
+  rw [mem_span_iff] at hx ⊢
+  exact fun I hI => hx I $ h.trans hI
+
+end NonUnitalNonAssocRing
+
+section NonAssocRing
+
+variable {R S : Type*} [NonAssocRing R] [NonAssocRing S]
+variable {F : Type*} [FunLike F R S] [RingHomClass F R S] (f : F)
+
+/--
+preimage of a two-sided-ideal is still two-sided -/
+def comap (I : TwoSidedIdeal S) : TwoSidedIdeal R :=
+{ ringCon := I.ringCon.comap f }
+
+lemma mem_comap {I : TwoSidedIdeal S} {x : R} :
+    x ∈ I.comap f ↔ f x ∈ I := by
+  simp [comap, RingCon.comap, mem_iff]
+
+/--
+pushout of a two-sided-ideal defined as the span of the image of a two-sided-ideal under some ring
+homomorphism.
+-/
+def map (I : TwoSidedIdeal R) : TwoSidedIdeal S :=
+  span (f '' I)
+
+lemma map_mono {I J : TwoSidedIdeal R} (h : I ≤ J) :
+    map f I ≤ map f J :=
+  span_mono $ Set.image_mono h
+
+/--
+kernel of a ring homomorphism as a two-sided-ideal.
+-/
+def ker (f : R →+* S) : TwoSidedIdeal R :=
+  .mk'
+    {r | f r = 0} (map_zero _) (by rintro _ _ (h1 : f _ = 0) (h2 : f _ = 0); simp [h1, h2])
+    (by rintro _ (h : f _ = 0); simp [h]) (by rintro _ _ (h : f _ = 0); simp [h])
+    (by rintro _ _ (h : f _ = 0); simp [h])
+
+lemma mem_ker (f : R →+* S) {x : R} : x ∈ ker f ↔ f x = 0 := by
+  delta ker; rw [mem_mk']; rfl
+  · rintro _ _ (h1 : f _ = 0) (h2 : f _ = 0); simp [h1, h2]
+  · rintro _ (h : f _ = 0); simp [h]
+  · rintro _ _ (h : f _ = 0); simp [h]
+  · rintro _ _ (h : f _ = 0); simp [h]
+
+end NonAssocRing
+
+section Ring
+
+variable {R : Type*} [Ring R]
+
+lemma mem_span_iff_exists {s : Set R} {x} :
+    x ∈ span s ↔
+    ∃ (ι : Type) (fin : Fintype ι) (xL : ι → R) (xR : ι → R) (y : ι → s),
+      x = ∑ i : ι, xL i * (y i : R) * xR i := by
```

Comment:
1. Why not work with `Finset` instead of the existence of a `Fintype`? It seems to me that would be far less hassle. 
2. This version is only true for `Ring`s. It would be nice to have the other statement for non-unital rings. (The one for non-associative rings I think is too messy to include, unless someone can convince me otherwise.)

Reply:
Re 2:
I am not sure I know how to prove it without a `1`. Say $x \in s$ and I want to prove that $x = \sum_{i \in I} x_i y_i x_i'$ for some indexing set $I$, some $x_i, x_i' \in R$ and $y_i \in I$ , my approach was $I$ is a one point set works and $x_i = x_i' = 1$ and $y_i = x$ works

Reply:
Re 1: I think you are thinking proving the following
```
example {ι : Type*} {s : Set R} {x} :
    x ∈ span s ↔
    ∃ (t : Finset ι) (xL : ι → R) (xR : ι → R) (y : ι → s),
      x = ∑ i ∈ t, xL i * (y i : R) * xR i
```

Assume we still use the same approach, i.e. prove `S := {x | ∃ (t : Finset ι) (xL : ι → R) (xR : ι → R) (y : ι → s),
      x = ∑ i ∈ t, xL i * (y i : R) * xR i}` is a two sided ideal. Then when proving `0` is in `S`, we need to find an actual `ι → s` but then we need to exclude the case where `s` is empty and prove it separately.

Reply:
Re 2: the statement *as written* is definitely not true without a 1. There is an analogous statement for non-unital rings. The ideal generated by a set consists of the sum of four sums: elements of the set, product, of elements of the set with elements of the ring, products of elements of the ring with elements of the set, and triple products of the form you wrote.

Reply:
Re 1: I would have maybe phrased it differently in the proof, maybe something like: `S` is the set of sums over `Finset`s where each term in the finset can be written as a triple product.

Reply:
Re 2: I have added a proof wanted, I tried to fiddle with it a bit, but couldn't do it. Can you check if it is the statement that you want:
```lean
proof_wanted mem_span_iff_exists_nonunital {s : Set R} {z : R} :
    z ∈ span s ↔
    ∃ (t : Finset R) (_ : (t : Set R) ⊆ s) (n : R → ℤ)
      (α : R → R) (β : R → R) (xy : R → R →₀ (R × R)),
      z = ∑ r ∈ t, (n r • r + α r * r + r * β r +
        ∑ i ∈ (xy r).support, (xy r i).1 * r * (xy r i).2)
```

Reply:
Re 1:
I could state it like the following
```lean
lemma mem_span_iff_exists' {s : Set R} {x} :
    x ∈ span s ↔
    ∃ (c : R →₀ R × R) (_ : (c.support : Set R) ⊆ s), x = c.sum fun i xy => xy.1 * i * xy.2 := by
```
But I struggled with the proof: when proving `{x | ∃ (c : R →₀ R × R) (_ : (c.support : Set R) ⊆ s), x = c.sum fun i xy => xy.1 * i * xy.2}` is a two sided ideal, I can't get the closed under addition working for it involves two `Finsupp` whose support are not disjoint.
By using `Fintype`, I can get disjointness for free.

Reply:
Re 1:
I am convinced the following is the best form:
```lean
lemma mem_span_iff_exists' {s : Set R} {x} :
    x ∈ span s ↔
    ∃ (c : R →₀ R × R) (_ : (c.support : Set R) ⊆ s), x = c.sum fun i xy => xy.1 * i * xy.2 := by
```
But I struggle again to prove it:
I want to say `{x | ∃ (c : R →₀ R × R) (_ : (c.support : Set R) ⊆ s), x = c.sum fun i xy => xy.1 * i * xy.2}` is a two-sided-ideal, but I can't prove that it is closed under addition

Reply:
okay, I will have a look. Thanks for thinking about this.

Reply:
Indeed, allowing `ℤ`-scalar multiples in the sum corrects this.

Reply:
@jjaassoonn I took the liberty of pushing a few results. They are not phrased exactly in the same way, but I think they should be reasonably useful as is.

In fact, by combining [AddSubgroup.mem_sup](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/Subgroup/Basic.html#AddSubgroup.mem_sup), [AddSubgroup.closure_union](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/Subgroup/Basic.html#AddSubgroup.closure_union), and [AddSubmonoid.exists_multiset_of_mem_closure](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/Submonoid/Membership.html#AddSubmonoid.exists_multiset_of_mem_closure), it is possible to recover from my results some expressions in terms of sums of multisets.

Reply:
Thanks a lot! I have removed `TwosidedIdeal.mem_span_iff_exists` for now in favour of your lemmas!